packages/core/src/browser/shell/tab-bar-decorator.ts

```diff
@@ -0,0 +1,88 @@
+/********************************************************************************
+ * Copyright (C) 2019 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, named, postConstruct } from 'inversify';
+import { Event, Emitter, Disposable, DisposableCollection, ContributionProvider } from '../../common';
+import { Title, Widget } from '@phosphor/widgets';
+import { WidgetDecoration } from '../widget-decoration';
+
+export const TabBarDecorator = Symbol('TabBarDecorator');
+
+export interface TabBarDecorator {
+
+    /**
+     * The unique identifier of the tab bar decorator.
+     */
+    readonly id: string;
+
+    /**
+     * Event that is fired when any of the available tabbar decorators has changes.
+     */
+    readonly onDidChangeDecorations: Event<void>;
+
+    /**
+     * Decorate tabs by the underlying URI.
+     * @returns A map from URI of the tab to its decoration data.
+     */
+    decorate(titles: Title<Widget>[]): Map<string, WidgetDecoration.Data>;
+}
+
+@injectable()
+export class TabBarDecoratorService implements Disposable {
+
+    protected readonly onDidChangeDecorationsEmitter = new Emitter<void>();
+
+    readonly onDidChangeDecorations = this.onDidChangeDecorationsEmitter.event;
+
+    protected readonly toDispose = new DisposableCollection();
+
+    @inject(ContributionProvider) @named(TabBarDecorator)
+    protected readonly contributions: ContributionProvider<TabBarDecorator>;
+
+    @postConstruct()
+    protected init(): void {
+        const decorators = this.contributions.getContributions();
+
+        decorators.forEach(decorator => {
```

Comment:
Do we not need to dispose these decorator listeners?
Looking at the `tree-decorator`, it disposes all of them:

https://github.com/theia-ide/theia/blob/c1d367f30333113423689af9344a4c7e8e15579d/packages/core/src/browser/tree/tree-decorator.ts#L119-L124

Reply:
I wasn't completely sure about this. But I guess there is no harm to dispose them?